astropy/units/structured.py

```diff
@@ -0,0 +1,465 @@
+        return result
+
+    def _recursively_get_dtype(self, value, enter_lists=True):
+        """Get structured dtype according to value, using our field names.
+
+        This is useful since ``np.array(value)`` would treat tuples as lower
+        levels of the array, rather than as elements of a structured array.
+        The routine does presume that the type of the first tuple is
+        representative of the rest.  Used in ``_get_converter``.
+
+        """
+        if enter_lists:
+            while isinstance(value, list):
+                value = value[0]
+        if not isinstance(value, tuple) or len(self) != len(value):
+            raise ValueError(f"cannot interpret value {value} for unit {self}.")
+        descr = []
+        for (name, unit), part in zip(self.items(), value):
+            if isinstance(unit, StructuredUnit):
+                descr.append(
+                    (name, unit._recursively_get_dtype(part, enter_lists=False)))
+            else:
+                # Got a part associated with a regular unit. Gets its dtype.
+                # Like for Quantity, we cast integers to float.
+                part = np.array(part)
+                part_dtype = part.dtype
+                if part_dtype.kind in 'iu':
+                    part_dtype = np.dtype(float)
+                descr.append((name, part_dtype, part.shape))
+        return np.dtype(descr)
+
+    @property
+    def si(self):
+        """The `StructuredUnit` instance in SI units."""
+        return self._recursively_apply(operator.attrgetter('si'))
+
+    @property
+    def cgs(self):
+        """The `StructuredUnit` instance in cgs units."""
+        return self._recursively_apply(operator.attrgetter('cgs'))
+
+    # Needed to pass through Unit initializer, so might as well use it.
+    def _get_physical_type_id(self):
+        return self._recursively_apply(
+            operator.methodcaller('_get_physical_type_id'), cls=Structure)
+
+    @property
+    def physical_type(self):
+        """Physical types of all the fields."""
+        return self._recursively_apply(
+            operator.attrgetter('physical_type'), cls=Structure)
+
+    def decompose(self, bases=set()):
+        """The `StructuredUnit` composed of only irreducible units.
+
+        Parameters
+        ----------
+        bases : sequence of `~astropy.units.UnitBase`, optional
+            The bases to decompose into.  When not provided,
+            decomposes down to any irreducible units.  When provided,
+            the decomposed result will only contain the given units.
+            This will raises a `UnitsError` if it's not possible
+            to do so.
+
+        Returns
+        -------
+        `~astropy.units.StructuredUnit`
+            With the unit for each field containing only irreducible units.
+        """
+        return self._recursively_apply(
+            operator.methodcaller('decompose', bases=bases))
+
+    def is_equivalent(self, other, equivalencies=[]):
```

Comment:
Would it be useful to have an element-wise option? maybe a flag.

Reply:
I'm not sure, what would be the use case?

I do wonder about allowing single units, i.e., that `u.Unit('cm,m,km').is_equivalent('m')` would be `True`.

Reply:
But if `is_equivalent` would be true, obviously `.to('m')` should work too, and in the other direction as well. I think that is better done in follow-up - and then decide too whether to deprecate structured arrays with single units.